### PR TITLE
Service resource methods are not considered for subtyping

### PR DESCRIPTION
**Pull request by rdhananjaya**
## Purpose
As a part of service typing stage-2, $title.

This is a breaking change, this disallow resource method decl in object type decl.
Example: https://github.com/ballerina-platform/ballerina-lang/pull/27410/files#diff-57bb99a9a4e0c7e4130933445e9369df2af89036d22e35d83082e976eb1a1713L33

## Check List 
- [ ] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

**Comment by MaryamZi** [comment]
Shall we change the base to https://github.com/ballerina-platform/ballerina-lang/tree/lang-sl-alpha?

File: compiler/ballerina-lang/src/main/resources/compiler.properties [hunk 1]
```diff
@@ -769,6 +769,12 @@ error.resource.function.invalid.return.type=\
 error.remote.in.non.object.function=\
   remote modifier not allowed in non-object attached function {0}
 
+error.unsupported.resource.function.declaration.in.object.type=\
+  unsupported resource function declaration in object type definition
+
+error.unsupported.remote.function.decl.in.service.obj.type=\
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/resources/compiler.properties:775]: Is this needed?

File: compiler/ballerina-lang/src/main/resources/compiler.properties [hunk 2]
```diff
@@ -769,6 +769,12 @@ error.resource.function.invalid.return.type=\
 error.remote.in.non.object.function=\
   remote modifier not allowed in non-object attached function {0}
 
+error.unsupported.resource.function.declaration.in.object.type=\
+  unsupported resource function declaration in object type definition
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/resources/compiler.properties:773]: Shall we use method instead?

```suggestion
  resource method declarations are not allowed in an object type definition
```

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/TypeChecker.java [hunk 3]
```diff
@@ -1639,6 +1639,12 @@ private static boolean checkObjectSubTypeForMethods(List<TypePair> unresolvedTyp
                                                         String targetTypeModule, String sourceTypeModule,
                                                         BObjectType sourceType, BObjectType targetType) {
         for (MemberFunctionType lhsFunc : targetFuncs) {
+            // As of state-2 of service typing changes, resource functions are not considered for sub-typing.
+            if (SymbolFlags.isFlagOn(targetType.getFlags(), SymbolFlags.SERVICE)
```
Comment (MaryamZi) [inline bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/TypeChecker.java:1643]: Do we need to check this? Can't we rely on just the resource flag for the method?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/BIRGen.java [hunk 4]
```diff
@@ -494,12 +494,17 @@ public void visit(BLangClassDefinition classDefinition) {
 
         for (BAttachedFunction func : ((BObjectTypeSymbol) classDefinition.symbol).referencedFunctions) {
             BInvokableSymbol funcSymbol = func.symbol;
+            if (Symbols.isService(classDefinition.symbol)
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/BIRGen.java:497]: Same comment as above. Since resources are only allowed in services do we need to check this here?

File: compiler/ballerina-lang/src/main/java/org/ballerinalang/util/diagnostic/DiagnosticErrorCode.java [hunk 5]
```diff
@@ -653,21 +653,24 @@
     REMOTE_FUNCTION_IN_NON_NETWORK_OBJECT("BCE3967", "remote.function.in.non.network.object"),
     UNSUPPORTED_PATH_PARAM_TYPE("BCE3968", "unsupported.path.param.type"),
     UNSUPPORTED_REST_PATH_PARAM_TYPE("BCE3969", "unsupported.rest.path.param.type"),
-    SERVICE_ABSOLUTE_PATH_OR_LITERAL_IS_REQUIRED_BY_LISTENER("BCE3970",
+    OBJECT_TYPE_DEF_DOES_NOT_ALLOW_RESOURCE_FUNC("BCE3970", "unsupported.resource.function.declaration.in.object.type"),
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/ballerinalang/util/diagnostic/DiagnosticErrorCode.java:656]: Shall we include the "declaration" part in the code?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java [hunk 6]
```diff
@@ -3532,6 +3528,11 @@ private void validateReferencedFunction(Location pos, BAttachedFunction func, Sy
             return;
         }
 
+        // Service typing does not consider resource methods when type checking.
+        if (Symbols.isService(func.symbol.receiverSymbol) && Symbols.isResource(func.symbol)) {
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java:3532]: Same comment as above re: checking for services first.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java [hunk 7]
```diff
@@ -1318,7 +1317,7 @@ private boolean checkStructEquivalency(BType rhsType, BType lhsType, Set<TypePai
         return false;
     }
 
-    public boolean checkObjectEquivalency(BObjectType rhsType, BObjectType lhsType, Set<TypePair> unresolvedTypes) {
+    public boolean  checkObjectEquivalency(BObjectType rhsType, BObjectType lhsType, Set<TypePair> unresolvedTypes) {
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java:1320]: Unnecessary change?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java [hunk 8]
```diff
@@ -3787,7 +3775,9 @@ private boolean checkDetachMethod(BAttachedFunction func) {
             }
 
             BType firstParamType = func.type.paramTypes.get(0);
-            return detachFound = isServiceObject(firstParamType);
+            boolean isMatchingSignature = firstParamType.tag == TypeTags.OBJECT
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java:3778]: Do we need a separate variable here?

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/TypeChecker.java (hunk 3 above)
Comment (rdhananjaya) [reply bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/TypeChecker.java:1643]: Yes we shold, will remove

File: compiler/ballerina-lang/src/main/java/org/ballerinalang/util/diagnostic/DiagnosticErrorCode.java [hunk 9]
```diff
@@ -653,23 +653,25 @@
     REMOTE_FUNCTION_IN_NON_NETWORK_OBJECT("BCE3967", "remote.function.in.non.network.object"),
     UNSUPPORTED_PATH_PARAM_TYPE("BCE3968", "unsupported.path.param.type"),
     UNSUPPORTED_REST_PATH_PARAM_TYPE("BCE3969", "unsupported.rest.path.param.type"),
-    SERVICE_ABSOLUTE_PATH_OR_LITERAL_IS_REQUIRED_BY_LISTENER("BCE3970",
+    OBJECT_TYPE_DEF_DOES_NOT_ALLOW_RESOURCE_FUNC_DECL("BCE3970",
+            "unsupported.resource.function.declaration.in.object.type"),
+    SERVICE_ABSOLUTE_PATH_OR_LITERAL_IS_REQUIRED_BY_LISTENER("BCE3972",
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/ballerinalang/util/diagnostic/DiagnosticErrorCode.java:658]: BCE3971? Need to change the rest too.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/BIRGen.java [hunk 10]
```diff
@@ -494,12 +494,15 @@ public void visit(BLangClassDefinition classDefinition) {
 
         for (BAttachedFunction func : ((BObjectTypeSymbol) classDefinition.symbol).referencedFunctions) {
             BInvokableSymbol funcSymbol = func.symbol;
+            if (Symbols.isResource(funcSymbol) && Symbols.isFunctionDeclaration(funcSymbol)) {
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/BIRGen.java:497]: In what scenarios is the included function a resource declaration?

I tried the following but `((BObjectTypeSymbol) classDefinition.symbol).referencedFunctions` is empty at this point.

```ballerina
service class Foo {
    resource function get x() returns int {
        return 1;
    }
}

service class Bar {
    *Foo;

    resource function get x() returns byte {
        return 1;
    }
}
```

Comment (rdhananjaya) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/BIRGen.java:497]: Oh yes, this doesn't happen as we disallow resources decl.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java [hunk 11]
```diff
@@ -1360,18 +1360,18 @@ public boolean checkObjectEquivalency(BObjectType rhsType, BObjectType lhsType,
...
-            }
         }
 
         for (BAttachedFunction lhsFunc : lhsFuncs) {
             if (lhsFunc == lhsStructSymbol.initializerFunc) {
                 continue;
             }
 
+            // Service resource methods are not considered as part of service objects type.
+            if (isLhsAService && Symbols.isResource(lhsFunc.symbol)) {
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java:1371]: Probably don't need the `isLhsAService` here also?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java [hunk 12]
```diff
@@ -3804,7 +3794,11 @@ private boolean checkAttachMethod(BAttachedFunction func) {
 
             // todo: change is unions are allowed as service type.
             BType firstParamType = func.type.paramTypes.get(0);
-            if (!isServiceObject(firstParamType)) {
+            if (firstParamType.tag != TypeTags.OBJECT) {
+                return false;
+            }
+
+            if (!Symbols.isService(firstParamType.tsymbol)) {
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java:3801]: Can't we continue to use `isServiceObject`?

**Review by MaryamZi — APPROVED** [review]
